lib/Parse/ParseStmt.cpp

```diff
@@ -1222,25 +1231,60 @@ validateAvailabilitySpecList(Parser &P,
     }
   }
 
-  if (ParsingMacroDefinition) {
-    if (HasOtherPlatformSpec) {
-      SourceLoc InsertWildcardLoc = P.PreviousLoc;
-      P.diagnose(InsertWildcardLoc, diag::attr_availability_wildcard_in_macro);
+  if (OtherPlatformSpecLoc != None) {
+    switch (Source) {
+    case Parser::AvailabilitySpecSource::Macro: {
+      SourceLoc Loc = OtherPlatformSpecLoc.getValue();
+      P.diagnose(Loc, diag::attr_availability_wildcard_in_macro);
+      break;
     }
-  } else if (!HasOtherPlatformSpec) {
-    SourceLoc InsertWildcardLoc = P.PreviousLoc;
-    P.diagnose(InsertWildcardLoc, diag::availability_query_wildcard_required)
+    default: 
+      break;
+    }
+  } else {
+    switch (Source) {
+    case Parser::AvailabilitySpecSource::Available: {
+      SourceLoc InsertWildcardLoc = P.PreviousLoc;
+      P.diagnose(InsertWildcardLoc, diag::availability_query_wildcard_required)
         .fixItInsertAfter(InsertWildcardLoc, ", *");
+      break;
+    }
+    case Parser::AvailabilitySpecSource::Unavailable: {
+      // Imply * in #unavailable statements
+      auto wldcrd = new (P.Context) OtherPlatformAvailabilitySpec(SourceLoc());
+      RecognizedSpecs.push_back(wldcrd);
```

Comment:
This is semantic thing, not in the source code. Could we move this out of `Parse`?
Also, Using `SourceLoc()` here means [`PoundAvailableInfo::getEndLoc()` may return invalid `SourceLoc`](https://github.com/apple/swift/blob/4cc1035ba22a4d0726e3a00f7dad75463d54e5de/lib/AST/Stmt.cpp#L295-L302), that means `PoundAvailableInfo::getSourceRange()` may hit assertion failure. Not sure this is actually reachable, but still.

Reply:
Done, I'm picking the location of the latest query now too.